dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java

```diff
@@ -67,23 +56,17 @@ private GroupCommParameterProvider(
   }
 
   @Override
-  public synchronized void push(final LayerParameter[] parameterGradients) {
-    // do not store the input if it is not valid
-    if (!(parameterGradients == null || parameterGradients.length == 0)) {
-      parameterGradientList.add(parameterGradients);
-    }
-
-    if (++pushCount >= batchSize) {
-      pushCount = 0;
-
-      try {
-        parameterGradientReduceSender.send(parameterGradientList);
-      } catch (final NetworkException e) {
-        throw new RuntimeException("NetworkException while trying to send reduce", e);
-      } catch (final InterruptedException e) {
-        throw new RuntimeException("InterruptedException while trying to send reduce", e);
+  public synchronized void push(final int batchSize, final LayerParameter[] parameterGradients) {
+    try {
+      if (batchSize == 0 || parameterGradients == null || parameterGradients.length == 0) {
+        parameterGradientReduceSender.send(Collections.<Pair<Integer, LayerParameter[]>>emptyList());
+      } else {
+        parameterGradientReduceSender.send(Collections.singletonList(new Pair<>(batchSize, parameterGradients)));
```

Comment:
The group comm version now uses `List<Pair>`s as Reduce values, but the compute tasks send lists that contain only one object. Instead of sending `List<Pair>`s, how about sending the inner `Pair`s directly? This way, the summation of parameter gradients can be done across reducers instead of only at the controller task.